docs/source/tutorials/InvertLensEquation.ipynb

```diff
@@ -219,10 +276,290 @@
+    "    ax.plot(y1, y2, color=\"r\", zorder=1)\n",
+    "S = raytrace(E1[..., 0], E1[..., 1])\n",
+    "S = torch.stack(S, dim=-1)\n",
+    "\n",
+    "# Identify triangles that contain the source plane point\n",
+    "locate1 = torch.vmap(caustics.lenses.func.triangle_contains, in_dims=(0, None))(S, s)\n",
+    "patches = []\n",
+    "for e, loc in zip(E1, locate1):\n",
+    "    patches.append(\n",
+    "        Polygon(\n",
+    "            e,\n",
+    "            fill=loc,\n",
+    "            alpha=0.4 if loc else 1,\n",
+    "            color=\"tab:green\" if loc else \"k\",\n",
+    "            linewidth=1,\n",
+    "        )\n",
+    "    )\n",
+    "p = PatchCollection(patches, match_original=True)\n",
+    "ax.add_collection(p)\n",
+    "ax.set_xlim([-fov / 1.9, fov / 1.9])\n",
+    "ax.set_ylim([-fov / 1.9, fov / 1.9])\n",
+    "ax.set_axis_off()\n",
+    "plt.show()\n",
+    "\n",
+    "# Get all the neighbors and upsample the triangles\n",
+    "E2 = E1[locate1]\n",
+    "E2 = torch.vmap(caustics.lenses.func.triangle_neighbors)(E2)\n",
+    "E2 = E2.reshape(-1, 3, 2)\n",
+    "E2 = caustics.lenses.func.remove_triangle_duplicates(E2)\n",
+    "# Upsample the triangles\n",
+    "E2 = torch.vmap(caustics.lenses.func.triangle_upsample)(E2)\n",
+    "E2 = E2.reshape(-1, 3, 2)\n",
+    "\n",
+    "fig, ax = plt.subplots()\n",
+    "CS = ax.contour(thx, thy, detA, levels=[0.0], colors=\"b\", zorder=1)\n",
+    "for path in paths:\n",
+    "    # Collect the path into a discrete set of points\n",
+    "    x1 = torch.tensor(list(float(vs[0]) for vs in path))\n",
+    "    x2 = torch.tensor(list(float(vs[1]) for vs in path))\n",
+    "    # raytrace the points to the source plane\n",
+    "    y1, y2 = lens.raytrace(x1, x2, z_s)\n",
+    "\n",
+    "    # Plot the caustic\n",
+    "    ax.plot(y1, y2, color=\"r\", zorder=1)\n",
+    "S = raytrace(E2[..., 0], E2[..., 1])\n",
+    "S = torch.stack(S, dim=-1)\n",
+    "\n",
+    "# Identify triangles that contain the source plane point\n",
+    "locate2 = torch.vmap(caustics.lenses.func.triangle_contains, in_dims=(0, None))(S, s)\n",
+    "patches = []\n",
+    "for e, loc in zip(E2, locate2):\n",
+    "    patches.append(\n",
+    "        Polygon(\n",
+    "            e,\n",
+    "            fill=loc,\n",
+    "            alpha=0.4 if loc else 1,\n",
+    "            color=\"tab:green\" if loc else \"k\",\n",
+    "            linewidth=0.5,\n",
+    "        )\n",
+    "    )\n",
+    "p = PatchCollection(patches, match_original=True)\n",
+    "ax.add_collection(p)\n",
+    "ax.set_xlim([-fov / 1.9, fov / 1.9])\n",
+    "ax.set_ylim([-fov / 1.9, fov / 1.9])\n",
+    "ax.set_axis_off()\n",
+    "plt.show()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "61fdd482",
+   "metadata": {},
+   "source": [
+    "The process repeats until the triangles have converged to a very small area, at which point we then run a root finding algorithm to get the final points."
```

Comment:
If I overlay the demagnified image found using the full `forward_raytrace` in the first section with the final images produced here, it looks as though they don't overlap. At first, I thought this was because the recursive iteration was stopped when it was still coarse. But it looks like you're running `forward_raytrace_rootfind` on the solutions to get the true points, so I'm not sure why they appear so offset. 
<img width="414" alt="image" src="https://github.com/user-attachments/assets/e92a787f-e9ec-4077-bf9d-15246075bf4f">

Reply:
Yeah, I'll have to look into that. My guess is that the root finder is getting stuck and ends up just returning the input point. But the full forward_raytrace algorithm does go much deeper in the recursion, so I think that one is very accurate. Still, I'll see what's going on with the root finder.